hlapipeline/alignimages.py

```diff
@@ -296,45 +318,77 @@ def perform_align(input_list, archive=False, clobber=False, update_hdr_wcs=False
                                 item.best_meta = item.meta.copy()
-                    # If a reasonable fit has been found, this is the initial setting of the best_fit_xxx variables
-                    else:
-                        if fit_rms < MAX_FIT_LIMIT:
+                            best_fitStatusDict = fitStatusDict.copy()
+                            break #break out of while loop
+                        elif fitQual < best_fitQual: # better solution found. keep looping but with the better solution as "best" for now.
+                            print("Better solution found!")
                             best_fit_rms = fit_rms
                             best_fit_num = fit_num
                             for item in imglist:
                                 item.best_meta = item.meta.copy()
-
+                            best_fitStatusDict = fitStatusDict.copy()
+                            best_fitQual = fitQual
+                        elif fitQual == best_fitQual: # new solution same level of fitQual. Choose whichever one has the lowest total rms as "best" and keep looping.
+                            if best_fit_rms >= 0.:
+                                if fit_rms < best_fit_rms:
+                                    best_fit_rms = fit_rms
+                                    best_fit_num = fit_num
+                                    for item in imglist:
+                                        item.best_meta = item.meta.copy()
+                                    best_fitStatusDict = fitStatusDict.copy()
+                        else: # new solution has worse fitQual. discard and continue looping.
+                            continue
                 except Exception:
                     print("WARNING: Catastrophic fitting failure with catalog {} and matching algorithm {}.".format(catalogList[catalogIndex],algorithm_name.__name__))
                     filteredTable['status'][:] = 1
                     # It may be there are additional catalogs and algorithms to try, so keep going
                     continue
 
-                # If this is true, we are done so break out of the catalog loop (inner for loop)
-                # THIS AND THE NEXT BREAK ARE FIXES TO THE CURRENT LOGIC.  IT IS ASSUMED THESE
-                # FIXES WILL BECOME OBSOLETE ONCE THE ENHANCED LOGIC FOR CHOOSING THE BEST
-                # SOLUTION IS IMPLEMENTED.
-                if best_fit_rms > 0 and best_fit_rms < MAX_FIT_RMS:
+                if fitQual == 1:  # break out of inner fit algorithm loop
                     break
-
-            # If this is true, we are done so break out of the catalog loop (outer for loop)
-            if best_fit_rms > 0 and best_fit_rms < MAX_FIT_RMS:
-                break
-
-                #imglist_temp = imglist.copy() # preserve best fit solution so that it can be inserted into a reinitialized imglist next time through.
-
+        if fitQual == 1: #break out of outer astrometric catalog loop
```

Comment:
Testing with I9ZL53010, the following exception was encountered:

```
Processing time of [STEP 5]: 0.334693 sec
Not enough sources found in catalog GAIADR2
Try again with other catalog
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
<ipython-input-21-69c9b055bb27> in <module>
----> 1 align_table = alignimages.perform_align(['I9ZL53010'], update_hdr_wcs=True)

/user/hack/dev/release/stsci_python/pypi/hlapipeline/git/hlapipeline/hlapipeline/alignimages.py in perform_align(input_list, archive, clobber, debug, update_hdr_wcs, print_fit_parameters, print_git_info)
    389                 if fitQual == 1:  # break out of inner fit algorithm loop
    390                     break
--> 391         if fitQual == 1: #break out of outer astrometric catalog loop
    392             break
    393     currentDT = datetime.datetime.now()

UnboundLocalError: local variable 'fitQual' referenced before assignment
```

Reply:
Issue resolved by commit [8fab63a](https://github.com/spacetelescope/hlapipeline/pull/73/commits/8fab63a16737d40bb385c1e79d9fe18f10e735c5)

Reply:
Issue resolved by commit [8fab63a](https://github.com/spacetelescope/hlapipeline/pull/73/commits/8fab63a16737d40bb385c1e79d9fe18f10e735c5)